{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA Assemblies, gene expression, transcription, and translation\n",
    "The central Dogma - namely that DNA is transcribed to mRNA which is translated to proteins - is a key part of modelling many biochemical processes, especially in synthetic biology. Towards enabling easy modeling of transcription and translation in diverse context, BioCRNpyler has a number of Mixtures, Components and Mechanisms to produce models which include the central dogma."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example 1: Creating a DNAassembly\n",
    "DNAassembly is a Component consisting of 2 sub Components: Promoter and RBS (ribosome binding site). DNAassembly automatically produces transcription and translation reactions from simple specifications: a DNAassembly X will produce 3 species dna_X, rna_X, and protein_X. These species can also be renamed manually if desired."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "repr(Mixture) gives a printout of what is in a mixture and what it's Mechanisms are:\n",
      " Mixture: Catalysis Mixture\n",
      "Components = [\n",
      "\tDNAassembly: X ]\n",
      "Mechanisms = {\n",
      "\ttranscription:simple_transcription\n",
      "\ttranslation:simple_translation } \n",
      "\n",
      "find parameter in Component\n",
      "self.mixture Mixture: Catalysis Mixture\n",
      "find parameter in mixture\n",
      "find parameter in Component\n",
      "self.mixture Mixture: Catalysis Mixture\n",
      "find parameter in mixture\n",
      "Pretty_print representation of the CRN:\n",
      " Species (3) = {0. dna[X], 1. protein[X], 2. rna[X]}\n",
      "Reactions (2) = [\n",
      "0. dna[X] --> dna[X] + rna[X]        \n",
      "        massaction: k_f(dna[X])=0.5*dna[X]\n",
      "1. rna[X] --> rna[X] + protein[X]        \n",
      "        massaction: k_f(rna[X])=2*rna[X]\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "from biocrnpyler import *\n",
    "\n",
    "#promoter is the name of the promoter or a Promoter component\n",
    "#RBS is the name of the RBS or an RBS Component\n",
    "#   RBS = None means there will be no translation\n",
    "#By default (if transcript and protein are None), the transgript and protein will have the same name as an assmebly\n",
    "#   Users can also assign Species or String names to transcript and protein\n",
    "G = DNAassembly(\"X\", promoter = \"P\", rbs = \"RBS\", transcript = None, protein = None)\n",
    "\n",
    "#create a transcription and translation Mechanisms. \n",
    "mech_tx = SimpleTranscription()\n",
    "mech_tl = SimpleTranslation()\n",
    "\n",
    "#place that mechanism in a dictionary: {\"transcription\":mech_tx, \"translation\":mech_tl}\n",
    "default_mechanisms = {mech_tx.mechanism_type:mech_tx, mech_tl.mechanism_type:mech_tl}\n",
    "\n",
    "#Use default parameters for conviencience\n",
    "default_parameters = {\"kb\":100, \"ku\":10, \"ktx\":.5, \"ktl\":2}\n",
    "#Create a mixture.\n",
    "M = Mixture(\"Catalysis Mixture\", components = [G], parameters = default_parameters, default_mechanisms = default_mechanisms)\n",
    "    \n",
    "print(\"repr(Mixture) gives a printout of what is in a mixture and what it's Mechanisms are:\\n\", repr(M),\"\\n\")\n",
    "\n",
    "#Compile the CRN with Mixture.compile_crn\n",
    "CRN = M.compile_crn()\n",
    "\n",
    "print(\"Pretty_print representation of the CRN:\\n\",\n",
    "      CRN.pretty_print(show_rates = True, show_attributes = True, show_materials = True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example 2: Placing a DNAassmebly in different Cell-like Mixtures for Transcription and Translation\n",
    "BioCRNpyler contains many pre-built Mixtures to model Transcription and Translation at different levels of detail in different contexts. Cell-like Mixtures are designed to more accurately model the internal environment of a cell.\n",
    "\n",
    "## Cell-Like Mixtures:\n",
    "__ExpressionDilutionMixture__ uses the Mechanisms: OneStepGeneExpression\n",
    "* Genes express in a single step $G \\to G + P_G$. \n",
    "* Proteins are degraded: $P_G \\to \\emptyset$.\n",
    "\n",
    "__SimpleTxTlDilutionMixture__ uses the Mechanisms: SimpleTranscription, SimpleTranslation, and Dilution (A global Mechanisms)\n",
    "* Genes transcribe via simple catalysis $G_X \\to G_X + T_X$\n",
    "* mRNAs translate via simple catalyis $T_X \\to T_X + P_X$. \n",
    "* Proteins and mRNA are degraded/diluted at different rates: $T_X \\to \\emptyset$, $P_X \\to \\emptyset$.\n",
    "\n",
    "__TxTlDilutionMixture__ uses the Mechanisms: Transcription_MM, Translation_MM, Degredation_mRNA_MM and Dilution (A global Mechanisms)\n",
    "* Genes transcribe via RNApolymerase (RNAP) $G_X + RNAP \\rightleftarrows G_X:RNAP \\to G_X + RNAP + T_X$\n",
    "* mRNAs translate via Ribosomes (Ribo) $T_X + Ribo \\rightleftarrows T_X:Ribo \\to T_X + Ribo + P_X$\n",
    "* mRNAs are degraded via endonucleases (Endo): $T_X + Endo \\rightleftarrows T_X:Endo \\to Endo$\n",
    "* Proteins & mRNA are Diluted: $P_X \\to \\emptyset$, $T_X \\to \\emptyset$\n",
    "* A Background DNAassembly $G_\\textrm{cellular_processes}$ puts loading on all the cellular resources.\n",
    "\n",
    "In the following examples, the parameter file default_parameters.txt will be used to quickly produce more complex models with realistic parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "find parameter in Component\n",
      "self.mixture ExpressionDilutionMixture: ExpressionDilution\n",
      "find parameter in mixture\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "[dna_X, protein_X] is not a Species! Can only set initial concentration of a Species.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-ea5adea0dd45>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;31m#Compare the Following Mixtures and Resulting CRNs\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[0mM1\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mExpressionDilutionMixture\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"ExpressionDilution\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcomponents\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mG\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparameter_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"default_parameters.txt\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m \u001b[0mCRN1\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mM1\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcompile_crn\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrepr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mM1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"\\n\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mCRN1\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpretty_print\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mshow_attributes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshow_material\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mshow_rates\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\"\\n\\n\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-0.2.1-py3.7.egg\\biocrnpyler\\mixtures_cell.py\u001b[0m in \u001b[0;36mcompile_crn\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     66\u001b[0m         \u001b[0mreactions\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[0mglobal_mech_reactions\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     67\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 68\u001b[1;33m         \u001b[0mspecies\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mset_initial_condition\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     69\u001b[0m         \u001b[0mspecies\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mlambda\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mrepr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     70\u001b[0m         \u001b[0mreactions\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mlambda\u001b[0m \u001b[0mr\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mrepr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mr\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-0.2.1-py3.7.egg\\biocrnpyler\\mixture.py\u001b[0m in \u001b[0;36mset_initial_condition\u001b[1;34m(self, s, component)\u001b[0m\n\u001b[0;32m    228\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mset_initial_condition\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcomponent\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    229\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mSpecies\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 230\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"{s} is not a Species! Can only set initial concentration of a Species.\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    231\u001b[0m         \u001b[0minit_conc\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    232\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mcomponent\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: [dna_X, protein_X] is not a Species! Can only set initial concentration of a Species."
     ]
    }
   ],
   "source": [
    "#Changing the promoter and RBS name will result in different parameters\n",
    "#Here are some parameters loaded into default_parameters.txt\n",
    "#Promoter Names: strong, medium, weak correspond to bicistronic RBSs: BCD2, BCD8 and BCD12\n",
    "#RBS Names: strong, medium, Weak correspond to Anderson Promoters: J23100, J23106, and J23103\n",
    "G = DNAassembly(\"X\", promoter = \"strong\", rbs = \"weak\", transcript = None, protein = None)\n",
    "\n",
    "#Compare the Following Mixtures and Resulting CRNs\n",
    "M1 = ExpressionDilutionMixture(\"ExpressionDilution\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN1 = M1.compile_crn()\n",
    "print(repr(M1),\"\\n\", CRN1.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n",
    "\n",
    "M2 = SimpleTxTlDilutionMixture(\"SimpleTxTl\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN2 = M2.compile_crn()\n",
    "print(repr(M2),\"\\n\", CRN2.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n",
    "\n",
    "M3 = TxTlDilutionMixture(\"e coli\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN3 = M3.compile_crn()\n",
    "print(repr(M3),\"\\n\", CRN3.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example 3: Placing a DNAassmebly in different Extract-like Mixtures for Transcription and Translation\n",
    "BioCRNpyler contains many pre-built Mixtures to model Transcription and Translation at different levels of detail in different contexts. Cell-like Mixtures are designed to more accurately model the internal environment of a cell.\n",
    "\n",
    "## Extract-Like Mixtures:\n",
    "__ExpressionDilutionMixture__ uses the Mechanisms: OneStepGeneExpression\n",
    "* Genes express in a single step $G \\to G + P_G$. \n",
    "__SimpleTxTlExtract__ uses the Mechanisms: SimpleTranscription, SimpleTranslation, and Dilution (A global Mechanisms)\n",
    "* Genes transcribe via simple catalysis $G_X \\to G_X + T_X$\n",
    "* mRNAs translate via simple catalyis $T_X \\to T_X + P_X$. \n",
    "* mRNA are degraded: $T_X \\to \\emptyset$\n",
    "\n",
    "__TxTlExtract__ uses the Mechanisms: Transcription_MM, Translation_MM, Degredation_mRNA_MM and Dilution (A global Mechanisms)\n",
    "* Genes transcribe via RNApolymerase (RNAP) $G_X + RNAP \\rightleftarrows G_X:RNAP \\to G_X + RNAP + T_X$\n",
    "* mRNAs translate via Ribosomes (Ribo) $T_X + Ribo \\rightleftarrows T_X:Ribo \\to T_X + Ribo + P_X$\n",
    "* mRNAs are degraded via endonucleases (Endo): $T_X + Endo \\rightleftarrows T_X:Endo \\to Endo$\n",
    "\n",
    "In the following examples, the parameter file default_parameters.txt will be used to quickly produce more complex models with realistic parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compare the Following Mixtures and Resulting CRNs\n",
    "M1 = ExpressionExtract(\"ExpressionExtract\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN1 = M1.compile_crn()\n",
    "print(repr(M1),\"\\n\", CRN1.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n",
    "\n",
    "M2 = SimpleTxTlExtract(\"SimpleTxTlExtract\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN2 = M2.compile_crn()\n",
    "print(repr(M2),\"\\n\", CRN2.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n",
    "\n",
    "M3 = TxTlExtract(\"e coli extract\", components = [G], parameter_file = \"default_parameters.txt\")\n",
    "CRN3 = M3.compile_crn()\n",
    "print(repr(M3),\"\\n\", CRN3.pretty_print(show_attributes = True, show_material = True, show_rates = True),\"\\n\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example 4: Retroactivity and Loading Using a Custom Promoter Object\n",
    "Most of the default Mixtures in BioCRNpyler include transcription, translation, and degredation machinery such as RNAP (RNA Polymerase), Ribosomes, and RNAases. In the following example, we will illustrate loading effects due to competition over this machinery. For this example, we will use the following model of constituitive transcription and translation:\n",
    "\n",
    "$G_i + \\textrm{RNAP} \\leftrightarrow G_i:\\textrm{RNAP} \\rightarrow G_i + \\textrm{RNAP} + T_i$\n",
    "\n",
    "$T_i + \\textrm{Ribosome} \\leftrightarrow T_i:\\textrm{Ribosome} \\rightarrow T_i + \\textrm{Ribosome} + P_i$\n",
    "\n",
    "$T_i + \\textrm{RNAase} \\leftrightarrow T_i:\\textrm{RNAase} \\rightarrow \\textrm{RNAase}$\n",
    "\n",
    "Here $G_i$, $T_i$ and $P_i$ are gene, transcript, and protein $i$, respectively. In the example that follows, we will allow different RNA polymerases for different genes. Also, some genes may not be translated at all. By using orthogonal polymerases and/or loads without translation, we will see different kinds of loading effects.\n",
    "\n",
    "We will create 4 different DNA assemblies. The reference assembly, \"ref\", will have a RNAP promoter and an RBS. We will examine the output of this reporter as a function of the amount of various load assemblies.\n",
    "* The \"Load\" assembly will be identical to the \"ref\" assembly; this assembly will put load on all parts of transcription, RNA degredation, and translation for the ref assembly.\n",
    "* The \"TxLoad\" assembly will have an RNAP promoter, but no RBS, ensuring that only RNAP and RNAases experience loading, not ribosomes.\n",
    "* The \"T7Load\" assembly will have a T7 promoter and an RBS so there will be no loading on polymerases.\n",
    "* The \"T7TxLoad\" assmebly will have a T7 promoter and no RBS, so there will only be load on the RNAases.\n",
    "\n",
    "The creation of these assemblies highlights the flexible, object oriented nature of bioCRNpyler."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biocrnpyler import *\n",
    "\n",
    "#Because we are lazy, all parameters will use the default \"param_name\"-->value key mapping.\n",
    "#Parameter warnings will be later suppressed in the Mixture constructor\n",
    "#For details on how parameter loading and defaulting works, see the Parameter notebook.\n",
    "kb, ku, ktx, ktl, kdeg = 100, 10, 3.0, 5.0, 2\n",
    "parameters = {\"kb\":kb, \"ku\":ku, \"ktx\":ktx, \"ktl\":ktl, \"kdeg\":kdeg}\n",
    "\n",
    "#A constituitively expressed reporter\n",
    "#By default the promoter 'P' will use the polymerase 'rnap'\n",
    "reference_assembly = DNAassembly(name = \"ref\", promoter = \"P\", rbs = \"BCD\")\n",
    "#A constiuitively expressed load (RNA and Protein)\n",
    "full_load_assembly = DNAassembly(name = \"Load\", promoter = \"P\", rbs = \"BCD\")\n",
    "#A constiutively transcribed (but not translated) load\n",
    "#By putting rbs = None, DNAassembly automatically knows not to include translation\n",
    "RNA_load_assembly = DNAassembly(name = \"TxLoad\", promoter = \"P\", rbs = None)\n",
    "\n",
    "#Load genes on orthogonal polymerases\n",
    "T7 = Protein(\"T7\") #Create a new protein (polymerase) called 'T7'\n",
    "\n",
    "#Create a custom promoter with a custom mechanism that uses T7 instead of RNAP\n",
    "#instantiate a new mechanism Transcription_MM with its own name and overwrote the default parameter rnap\n",
    "mechanism_txt7 = Transcription_MM(name = \"T7_transcription_mm\", rnap=T7)\n",
    "#Create an instance of a promoter with this mechanism for transcription\n",
    "T7P = Promoter(\"T7P\", mechanisms={\"transcription\":mechanism_txt7})\n",
    "#Create A load assembly with the custom T7 promoter\n",
    "T7_load_assembly = DNAassembly(name = \"T7Load\", promoter = T7P, rbs = \"BCD\")\n",
    "\n",
    "#Each new assembly requires its own promoter instance - so here I create another here\n",
    "T7P = Promoter(\"T7P\", mechanisms={\n",
    "    \"transcription\":Transcription_MM(name = \"T7_transcription_mm\", rnap=T7)})\n",
    "#A load assembly with the custom T7 promoter and no RBS\n",
    "T7RNA_load_assembly = DNAassembly(name = \"T7TxLoad\", promoter = T7P, rbs = None)\n",
    "\n",
    "#Add all the assemblies to a mixture\n",
    "components = [reference_assembly, full_load_assembly, T7_load_assembly, T7, RNA_load_assembly, T7RNA_load_assembly]\n",
    "myMixture = TxTlExtract(name = \"txtl\", parameters = parameters, components = components, parameter_warnings = False)\n",
    "\n",
    "#Print the CRN\n",
    "myCRN = myMixture.compile_crn()\n",
    "\n",
    "#The Species, Reaction, and CRN pretty_print functions return text which has been formated with a number of formatting options\n",
    "print(\"\\npretty_print gives a nicely formatted repesentation of the CRNS, reactions, and species. The names of species are formatted for clarity, but are not the actual species representations. Additionally a number of printing options are available.\",\n",
    "      \"\\n\", myCRN.pretty_print(show_material = True, show_rates = True, show_attributes = True))\n",
    "\n",
    "#Simulate with BioSCRAPE if installed\n",
    "\n",
    "print(\"Simulating\")\n",
    "import numpy as np\n",
    "import pylab as plt\n",
    "timepoints = np.arange(0, 3, .01)\n",
    "stochastic = False #Whether to use ODE models or Stochastic SSA models\n",
    "plt.figure(figsize = (16, 8))\n",
    "plt.subplot(221)\n",
    "plt.title(\"Load on a RNAP Promoter\")\n",
    "loads = [0, 1.0, 5., 10., 50, 100, 500, 1000]\n",
    "for dna_Load in loads:\n",
    "    #print(\"Simulating for dna_Load=\", dna_Load)\n",
    "    x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "               'dna_ref':5., 'dna_Load':dna_Load}\n",
    "\n",
    "    results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "    if results is not None:\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label = \"Load = \"+str(dna_Load))\n",
    "\n",
    "plt.xlim(0, 5)\n",
    "#plt.xlabel(\"time\")\n",
    "plt.ylabel(\"Reference Protein\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.title(\"Load on a T7 Promotoer\")\n",
    "for dna_Load in loads:\n",
    "    #print(\"Simulating for dna_T7Load=\", dna_Load)\n",
    "    x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "               'dna_ref':5., 'dna_T7Load':dna_Load}\n",
    "    results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "    if results is not None:\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "plt.xlim(0, 5)\n",
    "#plt.xlabel(\"time\")\n",
    "plt.ylabel(\"Reference Protein\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(223)\n",
    "plt.title(\"Load on a RNAP Promotoer, No RBS\")\n",
    "for dna_Load in loads:\n",
    "    #print(\"Simulating for dna_TxLoad=\", dna_Load)\n",
    "    x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "               'dna_ref':5., 'dna_TxLoad':dna_Load}\n",
    "    results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "    if results is not None:\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "plt.xlim(0, 5)\n",
    "plt.xlabel(\"time\")\n",
    "plt.ylabel(\"Reference Protein\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.title(\"Load on a T7 Promotoer, No RBS\")\n",
    "for dna_Load in loads:\n",
    "    #print(\"Simulating for dna_T7TxLoad=\", dna_Load)\n",
    "    x0_dict = {\"protein_T7\": 10., \"protein_RNAP\":10., \"protein_RNAase\":5.0, \"protein_Ribo\":50.,\n",
    "               'dna_ref':5., 'dna_T7TxLoad':dna_Load}\n",
    "    results = myCRN.simulate_with_bioscrape(timepoints, x0_dict, stochastic = stochastic)\n",
    "    if results is not None:\n",
    "        plt.plot(timepoints, results[\"protein_ref\"], label=\"Load = \" + str(dna_Load))\n",
    "plt.xlim(0, 5)\n",
    "plt.xlabel(\"time\")\n",
    "plt.ylabel(\"Reference Protein\")\n",
    "plt.legend()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
