{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA_construct Introduction\n",
    "DNA_construct is a flexible way to indicate a DNA sequence.\n",
    "\n",
    "A DNA sequence is composed of a string of nucleotides. The sequence of these nucleotides can allow binding of other biomolecules such as proteins or RNA. We choose to annotate known binding interactions and other features of DNA sequences by considering them to be \"DNA parts\". a \"part\" has a certain function, and the order and orientation of parts on a dna sequence describes the overall function of that DNA sequence.\n",
    "\n",
    "For example, a `Promoter` is a part which binds to RNA polymerase and causes everything downstream to be transcribed into RNA. A `Terminator`, then, is a sequence which stops the transcription of DNA into RNA. Therefore, anything between a `Promoter` and a `Terminator` gets turned into RNA.\n",
    "\n",
    "Other parts such as a Ribosome Binding Site or `RBS`, binds to the ribosome, but only if it is made of RNA and not DNA. Thus, an `RBS` is _like_ a `Promoter`, but instead of RNA it makes Protein. As always the position and orientation of parts determines what those parts actually do. If your sequence contains a Coding Sequence (`CDS`) _before_ an `RBS`, then the ribosome actually cannot translate that protein. Likewise if your `RBS` is after a `Terminator`, that `RBS` never gets turned into RNA.\n",
    "\n",
    "This type of logic is incorporated into `DNA_construct`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The basic workflow is to first define a list of parts, then string them together into a `DNA_construct`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "imports done!\n",
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? True\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? True\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? True\n",
      "\n",
      "[ptet_0, RNA_construct = UTR1_GFP_t16]\n",
      "construct update species\n",
      "\n",
      "CRN species\n",
      "dna[dna[ptet-forward]:dna[UTR1-forward]:dna[GFP-forward]:dna[t16-forward]]\n",
      "\n",
      "CRN reactions\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from biocrnpyler import *\n",
    "try:\n",
    "    #dnaplotlib is a cool library for plotting DNAs.\n",
    "    #please use my fork located at https://github.com/dr3y/dnaplotlib\n",
    "    #to install it type: pip install git+git://github.com/dr3y/dnaplotlib.git@master\n",
    "    import dnaplotlib as dpl\n",
    "    %matplotlib inline\n",
    "    dpl_enabled = True\n",
    "except (ModuleNotFoundError,ImportError) as e:\n",
    "    dpl_enabled = False\n",
    "print(\"imports done!\")\n",
    "\n",
    "def plotNetwork(inCRN,use_pretty_print=True,colordict = None,iterations=2000,rseed=30,posscale=1,export=False):\n",
    "    try:\n",
    "        from bokeh.models import (Plot , Range1d)\n",
    "\n",
    "        import bokeh.plotting\n",
    "        import bokeh.io\n",
    "        bokeh.io.output_notebook() #this makes the graph appear in line with the notebook\n",
    "        if(colordict is None):\n",
    "            colordict={\"complex\":\"cyan\",\"protein\":\"green\",\n",
    "                        \"GFP\":\"lightgreen\",\n",
    "                        \"RFP\":\"red\",\n",
    "                        \"ribosome\":\"blue\",\n",
    "                        \"dna\":\"white\",\"rna\":\"orange\",\n",
    "                        \"ligand\":\"pink\",\"phosphate\":\"yellow\",\"nothing\":\"purple\"}\n",
    "        layout_str = \"force\"\n",
    "        DG, DGspec, DGrxn = generate_networkx_graph(inCRN,use_pretty_print=use_pretty_print,colordict=colordict) #this creates the networkx objects\n",
    "        plot = Plot(plot_width=500, plot_height=500, x_range=Range1d(-500, 500), y_range=Range1d(-500, 500)) #this generates a \n",
    "        if(export):\n",
    "            plot.output_backend = \"svg\"\n",
    "        graphPlot(DG,DGspec,DGrxn,plot,layout=layout_str,posscale=posscale,iterations=iterations,rseed=rseed) #now you draw the network on the plot. Layout \"force\" is \n",
    "        bokeh.io.show(plot) #if you don't type this the plot won't show\n",
    "        if(export):\n",
    "            bokeh.io.export_svgs(plot, \"plot_file.svg\")\n",
    "            \n",
    "    except ModuleNotFoundError:\n",
    "        return None\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ptet = RegulatedPromoter(\"ptet\",[\"tetr\"],leak=True) #this is a repressible promoter\n",
    "pconst = Promoter(\"pconst\") #constitutive promoter\n",
    "utr1 = RBS(\"UTR1\") #regular RBS\n",
    "gfp = CDS(\"GFP\",\"GFP\")\n",
    "rfp = CDS(\"RFP\",\"RFP\")\n",
    "t16 = Terminator(\"t16\") #a terminator stops transcription\n",
    "\n",
    "\n",
    "#now that the parts are defined, we can put together our construct.\n",
    "parameters={\"cooperativity\":2,\"kb\":100, \"ku\":10, \"ktx\":.05, \"ktl\":.2, \"kdeg\":2,\"kint\":.05}\n",
    "\n",
    "construct_1 = DNA_construct([[ptet,\"forward\"],[utr1,\"forward\"],[gfp,\"forward\"],[t16,\"forward\"]],\\\n",
    "                                mechanisms = {\"transcription\":Transcription_MM(Species(\"RNAP\",material_type=\"protein\")),\\\n",
    "                                    \"translation\":Translation_MM(Species(\"Ribosome\",material_type=\"protein\")),\\\n",
    "                                        \"binding\":One_Step_Cooperative_Binding()},parameters=parameters) \n",
    "''\n",
    "w = construct_1.enumerate_components()\n",
    "print()\n",
    "print(w)\n",
    "\n",
    "x = construct_1.update_species()\n",
    "print()\n",
    "print(\"CRN species\")\n",
    "print('\\n'.join([a.pretty_print() for a in x]))\n",
    "\n",
    "y = construct_1.update_reactions()\n",
    "print()\n",
    "print(\"CRN reactions\")\n",
    "print('\\n'.join([a.pretty_print() for a in y]))\n",
    "#'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "imports done!\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARwAAABtCAYAAABz/d4wAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAQMklEQVR4nO3de3RV5ZnH8e8vJCFAuBhAkJMooBUQFaRqo1ilXUJFZJxaW7WKg0unHW29zLSr0zqjCIM6o3bsYta02qqtM8xUxamD2qrYcbx1ioAgKlDwhsQDIgrIzXBJnvlj7xNPIklOTpK990mez1pZK+c979n7OZf97He/e7/7lZnhnHNRKIo7AOdc9+EJxzkXGU84zrnIeMJxzkXGE45zLjKecJxzkelSCUfSryTNjTuOjiapl6THJH0saUHc8bSVAr+UtE3SkrjjaUrSVyXVSNol6YQExDNK0gpJOyVdE3c8HalLJZw4SbpJ0vw21J8p6cUcq58PDAEGmtnX2xqDJJN0lKRV4Ua1S1KdpNqsx9eHMdWFj3dIWinpnCbL+rmktZLqJc3MMf7TgMlApZmd3FJFSX0l/bOk9ZJ2S9og6WFJJ2fVsfC5TOzbw/JJYVy7wo11raTLcojvDuC7ZlZuZitaiG29pDOblLX4eUgaKenxMJ4PJd2WQzw/AJ41s75mNi+H+gXDE05hOAJYZ2YH2rMQMxsbblTlwAt8upGVm9ktYbU/hs8PAH4KPCBpQNZiVgJXAcvbGP96M9vdUiVJPYFngOOAc4B+wBjgAeDsJtXHZcWeHd/GMP5+wN8Cv5B0TA7xrcr1zTTR7OchqRR4muA9DQUqgVx2Su2JJ9nMrGD/gBMIvuidwIMEP8y5wCTgPeB7wAfAJuCyrNdNA1YAO4Aa4KYc1jUcMOBbwMZwmd8LnzsL2AfsB3YBK8Py/sC9Yd10GFsPgo2oFqgL629vYb2zmyz78hbq3gTMP0i5AUc1KXsWuKJJ2UzgxazHvcPXnnSQZb4IzMzhc7u8yXud3ULdK8LPqk8ry/zM+wnLJwHvNSnbApzfzHJ6hjEZsBt4q4V1/jtQD3wSvuYHrX0e4W/lhTb+pp8JP6vacD1Hx72ddeRfwbZwwr3HfxP8ECqABcDXsqoMJdjgUwQ/+n+VdEj43G7gUoK9+DTgSkl/nuOqvwR8DpgC/FDSmWb2JHAL8KAFe9xxYd37gQPAUQTJcQrBRr4G+CvC1oQ13kM3Ymazmiz73hzjbBdJPYDLCBLdu/kuJ4w3+73OaqH6mcBT1kpLKBeSiiR9leA7fq2Z2PZa0BqCoMV0ZHPLM7MZwAZgevg+cjk0qgbWS3oiPJx6VtJxLb3AzL5M49bnuhzWUzAKNuEQfJklwE/MbL+ZPQwszXp+PzAnfO53BHuLUQBm9qyZvWZm9Wb2KvBr4Iwc1zvbzHab2WvAL4GLDlZJ0hBgKnBdWP8D4E7gwra/1UhVh30itQR9G5eEsUdhEPB+5oGk8ZK2h/1Ja5vUXR4+t11Sdj/HsDD+D4FZwAwza/raqFQSfN/zgGHAb4GF4c6yWyrkhDMMSFvYDg1l74k/ssZ9HnuAcgBJX5D0v5K2SPqYYA88KMf11jRZ37Bm6h1BkBA3ZTYM4G7g0BzXk48D4TobSMo83p/jMhaHLa5DgEeBL3ZYdK37CDgs88DMXgljOY/g8CfbBDMbEP5ln8nZGJZVmNl4M3ug88Nu1icEh6hPmNk+ggQ+kOCQulsq5ISzCUhJUlbZ4Tm+9j8JNqYqM+sP3AWo5Zc0qGqyvo3h/02H3dcAe4FBWRtGPzMb20z9jrCBoK8p2wiCPoF0WxZkZrsIOkNnRHiq+H+AKZL6RLS+tmrrd/ZqHq/p0go54fyRYI9+jaRiSecBLZ5yzdIX2GpmteHp1m+2Yb03SOotaSxBH8eDYflmYLikIgAz2wQsAn4sqV/Yp3CkpDOy6ld2cPP6SWCUpBmSSiRVEPT/PGx5nOEys4+Ae4AbM2WSSiWVESToEkllmffcAf6NYEfyiKRjJfUI13ViBy2/vTYDI7MLWvk85hMcop4Z9oldR3CotybCmBOlYBNO2EQ9j+DMyjbgAuA3Ob78KmCOpJ0EG9NDbVj1c8CbBHvjO8xsUVieuSDvI0mZU6SXAqXA6jDGh/n0kOEZglOf70v6sA3rb1bY13I28G2Cs3OvAx8DV7ZjsT8BzpZ0fPh4EcGhwqnAz8P/T2/H8huYWS1Bp/xqgv6OHcBa4CTgGx2xjna6Ffj78BD5+2FZs59H2Hd0CUELehtwLvBn4W+3W1LjLhDXHEnDgXeAknxaC865Am7hOOcKjyecLJIuzrpcPvuv06/6bDLsIPvv4ib17mqm3l2dHWNLwmtNDhbX9U3qXd9MvSc6MbacvldJhzdTb5ekXE9I5BJPzuvJ9XdRKPyQyjkXGW/hOOci4wnHORcZTzjOuch4wnHORcYTjnMuMp5wnHOR8YTjnItMcdwBZKRTVacT3NJhQSpdUxt3PM65jpekFs4IgtHC76VTVbenU1XN3n3NOVeYkpRwMgYC3wfeTKeqnkqnqs5Np6oS0xJzzuUviQkn2xSC+xa/k05V3ZBOVR3WSn3nXIIlPeFkVAJzgA3pVNWCdKrqS+lUVa536HPOJUShJJyMYoJJ4Z4BVqdTVdemU1UD4g3JOZerQks42UYT3I1uYzpVdW86VfX5mONxzrWi3benSKeqknR/i6XAral0zSNxB+K6JknFBBMO3mAJubfL3efOjzyOby+8JK8ujUJu4TS1nGAalkWtVXSunf6OYGLFHnEHUmgK/XRzLcH0vj8DlqbSNYnY47hu4UqgQtKl3fmm6G3VoYdUqXRN3meO0qmqvwB+lWP1dQR3wr8/la7Zmu86nWur8JAqe1LBp4CvdcT0xFHKPgzL9/AoH4XUwqkjuCbnZ8Az3ppxCfEV4PeSppmZ7/xaUQgJJ00w3889qXTNxtYqOxeDauA5SV8xM/+NtiDJCedpgtbMY6l0jc8D5ZLuWOAPkiab2ZtxB5NUSTtLtRX4MXB0Kl0zJZWuecSTjUsqFYkTrhuHejR0gQwHXpQ0Lr6oki1JLZwngYdS6ZpP4g7EuVxVnpGipE8Jy25fTv2+eoAhBIdX083shZjDS5zEtHBS6ZrNnmxcIRpy4qFUzzqZ4t4N++/+wCJJ02IMK5ESk3CcK2QDj6ng1LnV9BxQmikqAxYW6gyZncUTjnMdpP+Ifky85RR6D+mVKeoBzJd0dYxhJUqS+nBcMySVAEcCxwAjgV6ACK60fhdYDawzs70duM5Lw3W5xlrcSfc5rA8TbzmFxbOXsHPDrkzxPEkDgdlJGX8VF084CSVpEHAecAHwRaCklZfUSVoKPAgsMLN0O0OYAZzZzmV0S2UVZZx68yksmbuUbWu3Z4pnAQMlXWtm9fFFFy8/pEoYScMlzQfeJxiM+mVaTzYQNN+rgTuBGkmPShrbeZG6lpSWl1B908kMPmFwdvF3CQ6xcvk+uyRv4SSEpHKC2x5cDZQ2fb603yB6DT6CsoGV9CjtBRL1+2qp3Zrmky0b2Lv9/UaLA6YD0yTdB1xvZlvyjS11xjD6DO2d78u7LKnlIUjFZcWc/KPPs2LeSja+uClTfBEwQNL5Zrans2NMGk84CSBpNPAbYEx2eXnlGAaOPYOKMadR2m9Qi8vYv2cH2/70Bz5a9Rw73lkJGAQt2CuAs8If+Ev5xFc1KcXg8YNbr+g+o6ikiAnXjaekvIR3n9yQKZ5KcNp8upltizG8yHnCiZmkicBjwCGZsvLUaA6ffAV9Dz825+WU9O7HoROmcuiEqezZ/DYbfn8fH7+1LPN0JfCspAvNbGFHxu9apx7iuG+NpbRvKW8saBj1MJFPx19tauHlXUq3TDiSOuVMgZm1aZi/pKOAx4EBAEXFPRl+9ncYNG5yq831lvQeMpLRF89l29rFvLXwDupqd0FwXciDkk43syV5L9zlRRKjv3k0pX1LWHXfmkzxcQRDISab2dsxhhcZ7zSOiaS+wELCZFPcZwBjZt7O4PFT2pVssh0yqppjr5hHz0MaZtfpCTwiyafbicnI6SMYf+04VNTwHY8kGPR5fIxhRcYTTnz+keC6GtSjhFEXzqZ82NEdvpKyimGMvvhmepSVZ4qGEZz9cjGpmpTixB9OoKi0YfMbSnB4NTHGsCLR7ROOmbXrLx+SjgD+MvN4xDnXUJ4a1VFv6TPKKobxufOvJzh5BcB0SSd12gpdq4aeNITqGxuNvxoAPC1panxRdb5un3Bi8iPCa2vKK49h0PGdf31d/5ETqBh7enbRrE5fqWvRwLEVnPoPX6C0f8NVEL2ARyVdFGNYncoTTsQkFRFM5gdA5aRLOqzPpjWVp19MVitnatiP5GLUf2R/Jt5yCr0GN4y/Kgb+Q9JVMYbVaTzhRG8MMBCguHd/+o04IbIV9xp8OL2HNgyPKgJOjWzlrlnlw/ow8dZTKK9q6GcTwTQ0NyqqvVFEuuVp8ZidlvmnrnY329fldS1e3upqG00ucBrBrAMuZr0GljFxbjUv3byM7eu2Z4pnE4y/+uuuMv4qp2liOuu6lSRo7+DdLrYDOqjqWSf5lcYROfDJAZb+03I+XPlh3KHkpbVr0fyQyrkEKe5VTOq0rnuZlB9SOZcgbz++nlX3ro47jE6TU8Jp6yX7SZeUoQ1JJulp/H44kTEz1j3wBuseajTDzGtAlxpr5S0c52Jm9cbrv1jF+k9HkwP8H3BOVxtN7gnHuRjV769ver8cCKZMOr/Q5ivPhScc52JyoPYAy25bwZYVje6N9mtgppntiymsTuUJx7kY7Nu1nyU3L2PbnxodMf0UuLqrXHNzMN0+4XSH62hcstRurW06qwPAHOCmrj6rQ7dPOM5Fafem3SyevYQ9mxtNMnutmc2LK6YoecJxLiI71u9g8Zyl7N3WMH1YHUF/zfwYw4pUt0w4Xel6GVcYtq7Zyktzl3Fgz4FMUS3wdTN7PMawItctE45rm30791O7tTbuMBKprKKs1TqbX/6AZbctp35fQ1/wDmC6mT3fmbElUU6DN13341cat05F4pz/avkGfe89n+aVea9idQ3b2QcEVw+/0snhJZK3cJzrJO/8bj2v37M6nCIMCOaBn2xmb8QXVbw84bjmbAW6zBieDiSCm543y8x446E3WftAo7yyiqBl09453wuaJxx3UGZ2QdwxJJGkYmB/c89bvbHqvtW889t3s4sXA9PMbGtnx5d0nnCc6yD1B+p55V9eJf38xuziRcB5XXFcVD484TjXAQ7srePl25fzwcuNxkU9BMzoquOi8uEJx7l22r87GBe1dU2jcVF3A98xs7qYwkokTzjOtUPttr28NGcJO9bvzC6+Gbihq4+Lyodfh+NcGzTqNBb0HtKbPe/vya7yN2Z2ZxyxFQJPOM61QQtnqeqAy83s/ohDKih+SOVc++0FvmFmj8YdSNJ5wnGufXYSjIt6Lu5ACoEnHOfytwU4y8yWxx1IofCE41x+NhCMi1oXdyCFxGfedK7t1gATPdm0nZ+lcq4NJBUBFWZWmJN/x8wTjnMuMn5I5ZyLjCcc51xkPOE45yLjCcc5FxlPOM65yHjCcc5FxhOOcy4ynnCcc5HxhOOci4wnHOdcZDzhOOci4wnHORcZTzjOuch4wnHORcYTjnMuMp5wnHOR8YTjnIuMJxznXGT+H2jxukaQ0W74AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 270x79.2 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "ptet_0 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? True\n",
      "UTR1_1 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "GFP_2 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "t16_3 forward\n",
      "effective direction forward\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 1\n",
      "am I a promoter? True\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 2\n",
      "am I a promoter? True\n",
      "t16_3 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "GFP_2 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "UTR1_1 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? False\n",
      "ptet_0 reverse\n",
      "effective direction reverse\n",
      "loop count 3\n",
      "am I a promoter? True\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Unable to find mechanism of type transcription in Component ptet_0 or Mixture None.'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-546eeb028614>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     33\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     34\u001b[0m \u001b[1;32mif\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdpl_enabled\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 35\u001b[1;33m     \u001b[0mplotConstruct\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mconstruct_1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdebug\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mplot_rnas\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     36\u001b[0m     \u001b[0mplotConstruct\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mconstruct_2\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mdebug\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mplot_rnas\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     37\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mconstruct_1\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-1.0.2-py3.7.egg\\biocrnpyler\\plotting.py\u001b[0m in \u001b[0;36mplotConstruct\u001b[1;34m(DNA_construct_obj, dna_renderer, rna_renderer, plot_rnas, debug, showlabels, plot_dna_test)\u001b[0m\n\u001b[0;32m    486\u001b[0m         \u001b[0mplotDesign\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdesign\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mcircular\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mcircular\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mtitle\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mDNA_construct_obj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_species\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    487\u001b[0m         \u001b[1;32mif\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mplot_rnas\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 488\u001b[1;33m             \u001b[0mrnas_and_proteins\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mDNA_construct_obj\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0menumerate_components\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    489\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0mcomponent\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrnas_and_proteins\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    490\u001b[0m                 \u001b[1;32mif\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcomponent\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_species\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmaterial_type\u001b[0m\u001b[1;33m==\u001b[0m\u001b[1;34m\"rna\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-1.0.2-py3.7.egg\\biocrnpyler\\dna_construct.py\u001b[0m in \u001b[0;36menumerate_components\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    777\u001b[0m                 \u001b[1;32mif\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mupdated_components\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    778\u001b[0m                     \u001b[0mactive_components\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mupdated_components\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 779\u001b[1;33m         \u001b[0mcombinatorial_complexes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate_combinatorial_complexes\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mactive_components\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    780\u001b[0m         \u001b[0mcombinatorial_components\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    781\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mcomb_specie\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mcombinatorial_complexes\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-1.0.2-py3.7.egg\\biocrnpyler\\dna_construct.py\u001b[0m in \u001b[0;36mupdate_combinatorial_complexes\u001b[1;34m(self, active_components)\u001b[0m\n\u001b[0;32m    713\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mpart\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mactive_components\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    714\u001b[0m             \u001b[1;31m#first we make binary complexes\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 715\u001b[1;33m             \u001b[0msp_list\u001b[0m \u001b[1;33m=\u001b[0m  \u001b[0mpart\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate_species\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    716\u001b[0m             \u001b[0mspecies\u001b[0m\u001b[1;33m+=\u001b[0m\u001b[0mremove_bindloc\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msp_list\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    717\u001b[0m         \u001b[1;31m#print(\"initial species\")\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-1.0.2-py3.7.egg\\biocrnpyler\\dna_part_promoter.py\u001b[0m in \u001b[0;36mupdate_species\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    138\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    139\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mupdate_species\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 140\u001b[1;33m         \u001b[0mmech_tx\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_mechanism\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"transcription\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    141\u001b[0m         \u001b[0mmech_b\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_mechanism\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'binding'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    142\u001b[0m         \u001b[0mspecies\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mC:\\ProgramData\\Anaconda3\\lib\\site-packages\\biocrnpyler-1.0.2-py3.7.egg\\biocrnpyler\\component.py\u001b[0m in \u001b[0;36mget_mechanism\u001b[1;34m(self, mechanism_type, optional_mechanism)\u001b[0m\n\u001b[0;32m    174\u001b[0m             \u001b[0mmech\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmixture\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_mechanism\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmechanism_type\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    175\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mmech\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m \u001b[1;32mand\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0moptional_mechanism\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 176\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34mf\"Unable to find mechanism of type {mechanism_type} in Component {self} or Mixture {self.mixture}.\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    177\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    178\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mmech\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Unable to find mechanism of type transcription in Component ptet_0 or Mixture None.'"
     ]
    }
   ],
   "source": [
    "from biocrnpyler import *\n",
    "try:\n",
    "    #dnaplotlib is a cool library for plotting DNAs.\n",
    "    #please use my fork located at https://github.com/dr3y/dnaplotlib\n",
    "    #to install it type: pip install git+git://github.com/dr3y/dnaplotlib.git@master\n",
    "    import dnaplotlib as dpl\n",
    "    dpl_enabled = True\n",
    "except (ModuleNotFoundError,ImportError) as e:\n",
    "    dpl_enabled = False\n",
    "\n",
    "print(\"imports done!\")\n",
    "\n",
    "#part definitions below\n",
    "\n",
    "ptet = RegulatedPromoter(\"ptet\",[\"tetr\"],leak=True) #this is a repressible promoter\n",
    "pconst = Promoter(\"pconst\") #constitutive promoter\n",
    "pcomb = CombinatorialPromoter(\"pcomb\",[\"arac\",\"laci\"], leak=False,\n",
    "                              tx_capable_list = [[\"arac\"], [\"laci\"]]) #the Combinations A and B or just A or just B be transcribed\n",
    "utr1 = RBS(\"UTR1\") #regular RBS\n",
    "utr2 = RBS(\"UTR1\") #regular RBS\n",
    "gfp = CDS(\"GFP\",\"GFP\") #a CDS has a name and a protein name. so this one is called GFP and the protein is also called GFP\n",
    "fusrfp = CDS(\"fusRFP\",\"RFP\",no_stop_codons=[\"forward\"]) #you can say that a protein has no stop codon. This is a little different from a fusion protein, because in this case you are saying that the ribosome reads through two proteins but still produces two distinct proteins, rather than one fused protein. This can happen in the case of the ta peptide which causes a peptide bond not to be formed while making a protein.\n",
    "rfp = CDS(\"RFP\",\"RFP\") #regular RFP\n",
    "cfp = CDS(\"CFP\",\"CFP\") #cfp\n",
    "t16 = Terminator(\"t16\") #a terminator stops transcription\n",
    "\n",
    "\n",
    "#now that the parts are defined, we can put together our construct.\n",
    "construct_1 = DNA_construct([[ptet,\"forward\"],[utr1,\"forward\"],[gfp,\"forward\"],[t16,\"forward\"]])\n",
    "construct_2 = DNA_construct([[t16,\"reverse\"],[rfp,\"reverse\"],[utr2,\"reverse\"],[pconst,\"reverse\"]])\n",
    "#now, we are using dnaplotlib to plot the constructs\n",
    "\n",
    "\n",
    "if(dpl_enabled):\n",
    "    plotConstruct(construct_1,debug=False,plot_rnas=True)\n",
    "    plotConstruct(construct_2,debug=False,plot_rnas=True)\n",
    "    print(construct_1.show())\n",
    "    print(construct_2.show())\n",
    "\n",
    "#some very basic parameters are defined\n",
    "parameters={\"cooperativity\":2,\"kb\":100, \"ku\":10, \"ktx\":.05, \"ktl\":.2, \"kdeg\":2,\"kint\":.05}\n",
    "\n",
    "components = [construct_1,construct_2]\n",
    "myMixture = TxTlExtract(name = \"txtl\", parameters = parameters, components = components)\n",
    "myCRN = myMixture.compile_crn()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can see that for the first construct, your DNA is given an automatic name: `ptet_f_UTR1_f_GFP_f_t16_f`. \n",
    "\n",
    "The RNA made from your DNA is also given an automatic name: `UTR1_f_GFP_f_t16_f`\n",
    "\n",
    "Likewise for the second construct, which I've made in reverse to make the difference more apparent. Though you'll see that the RNA from the second construct is still produced in the right orientation, so the RNA is shown in the forwards direction.\n",
    "\n",
    "This is because DNA and RNA are now represented as an OrderedPolymer. Each member of the OrderedPolymer represents a part. So, in the case of promoters or RBSes, you'll see ribosomes and polymerases bind directly to the part inside the OrderedPolymer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#here we are plotting the network diagram and simulating the system with some basic parameters\n",
    "\n",
    "plotNetwork(myCRN)\n",
    "try:\n",
    "    import numpy as np\n",
    "    import matplotlib.pyplot as plt\n",
    "    timepoints = np.linspace(0, 200, 1000)\n",
    "    x0 = {construct_1.get_species():5.0,\n",
    "        construct_2.get_species():2.0, \"protein_RNAP\":10., \"protein_Ribo\":50.}\n",
    "    Re1 = myCRN.simulate_with_bioscrape_via_sbml(timepoints, initial_condition_dict = x0)\n",
    "    if(Re1 is not None):\n",
    "        plt.plot(timepoints,Re1[\"protein_GFP\"], label = \"protein_GFP\")\n",
    "        plt.plot(timepoints,Re1[\"protein_RFP\"], label = \"protein_RFP\")\n",
    "        plt.title(\"Time trace of two DNA system\")\n",
    "        plt.xlabel(\"time\")\n",
    "        plt.ylabel(\"protein\")\n",
    "        plt.legend()\n",
    "except ModuleNotFoundError:\n",
    "    print('please install the plotting libraries: pip install biocrnpyler[all]')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That seems pretty straightforward but now let's see what happens if we put both on the same piece of DNA:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "construct_3 = DNA_construct([[ptet,\"forward\"],[utr1,\"forward\"],[gfp,\"forward\"],[t16,\"forward\"],[t16,\"reverse\"],[rfp,\"reverse\"],[utr1,\"reverse\"],[pconst,\"reverse\"]])\n",
    "#now, we are using dnaplotlib to plot the constructs\n",
    "\n",
    "if(dpl_enabled):\n",
    "    plotConstruct(construct_3,debug=False,plot_rnas=True)\n",
    "    print(construct_3.show())\n",
    "\n",
    "#some very basic parameters are defined\n",
    "parameters={\"cooperativity\":2,\"kb\":100, \"ku\":10, \"ktx\":.05, \"ktl\":.2, \"kdeg\":2,\"kint\":.05}\n",
    "\n",
    "components = [construct_3]\n",
    "myMixture = TxTlExtract(name = \"txtl\", parameters = parameters, components = components)\n",
    "myCRN = myMixture.compile_crn()\n",
    "''\n",
    "try:\n",
    "    import numpy as np\n",
    "    import matplotlib.pyplot as plt\n",
    "    timepoints = np.linspace(0, 200, 1000)\n",
    "    x0 = {construct_3.get_species():5.0, \"protein_GFP\":20,\"protein_RNAP\":10., \"protein_Ribo\":50.,\"protein_arac\":20}\n",
    "    #I started the GFP at a slightly elevated value so you can see the two lines on top of each other in the graph\n",
    "    Re1 = myCRN.simulate_with_bioscrape_via_sbml(timepoints, initial_condition_dict = x0)\n",
    "    if(Re1 is not None):\n",
    "        plt.plot(timepoints,Re1[\"protein_GFP\"], label = \"protein_GFP\")\n",
    "        plt.plot(timepoints,Re1[\"protein_RFP\"], label = \"protein_RFP\")\n",
    "        #plt.plot(timepoints,Re1[\"rna_mydna\"], label = \"rna_mydna\")\n",
    "        #plt.gca().set_yscale(\"log\")\n",
    "        plt.title(\"Time trace of single DNA system\")\n",
    "        plt.xlabel(\"time\")\n",
    "        plt.ylabel(\"protein\")\n",
    "        plt.legend()\n",
    "        \n",
    "except ModuleNotFoundError:\n",
    "    print('please install the plotting libraries: pip install biocrnpyler[all]')\n",
    "#'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The same RNAs and proteins are made, but the network diagram should be a bit more complicated, owing to the combinatorial binding to the DNA molecule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plotNetwork(myCRN)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can see that many more arrows lead to each RNA. But there is still only one promoter per RNA, how can it be? Well, now that both promoters and transcripts are on the same DNA, we must take into account the fact that you can have both promoters with RNA polymerase bound, so for example Promoter1 can lead to RNA1 whether or not Promoter2 is bound by RNAP.\n",
    "\n",
    "We can even make a more complicated DNA construct, such as what happens before and after a promoter is \"flipped\" by integrase activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "attB = AttachmentSite(\"attB\",\"attB\",integrase=\"Bxb1\")\n",
    "attP = AttachmentSite(\"attP\",\"attP\",integrase=\"Bxb1\")\n",
    "attL = AttachmentSite(\"attL\",\"attL\",integrase=\"Bxb1\")\n",
    "attR = AttachmentSite(\"attR\",\"attR\",integrase=\"Bxb1\")\n",
    "\n",
    "\n",
    "construct_1 = DNA_construct([[t16,\"reverse\"],[rfp,\"reverse\"],[utr1,\"reverse\"],[attP,\"forward\"],[pconst,\"forward\"],[attB,\"reverse\"],[utr1,\"forward\"],[gfp,\"forward\"],[t16,\"forward\"]])\n",
    "construct_2 = DNA_construct([[t16,\"reverse\"],[rfp,\"reverse\"],[utr1,\"reverse\"],[attR,\"forward\"],[pconst,\"reverse\"],[attL,\"reverse\"],[utr1,\"forward\"],[gfp,\"forward\"],[t16,\"forward\"]])\n",
    "#now, we are using dnaplotlib to plot the constructs\n",
    "\n",
    "if(dpl_enabled):\n",
    "    plotConstruct(construct_1,debug=False,plot_rnas=True)\n",
    "    plotConstruct(construct_2,debug=False,plot_rnas=True)\n",
    "    plotConstruct(construct_1,debug=False,plot_rnas=True,plot_dna_test=False)\n",
    "    plotConstruct(construct_2,debug=False,plot_rnas=True,plot_dna_test=False)\n",
    "\n",
    "#some very basic parameters are defined\n",
    "parameters={\"cooperativity\":2,\"kb\":100, \"ku\":10, \"ktx\":.05, \"ktl\":.2, \"kdeg\":2,\"kint\":.05}\n",
    "\n",
    "components = [construct_1,construct_2]\n",
    "myMixture = TxTlExtract(name = \"txtl\", parameters = parameters, components = components)\n",
    "myCRN = myMixture.compile_crn()\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And we can see that this construct 'knows' that a forward facing promoter will not produce RNAs that face the reverse direction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "plotNetwork(myCRN)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview for developers\n",
    "Several key additions had to be made to biocrnpyler to make this possible. \n",
    "\n",
    "1. `Complex()` is a new function that you should use every time you want to make a ComplexSpecies. It should do exactly the same thing as the ComplexSpecies constructor, unless one of the species that you gave it is an OrderedComplexSpecies that has a 'bindloc' attribute, then instead of making a ComplexSpecies the function will make an OrderedComplexSpecies that is correctly formed.\n",
    "\n",
    "2. `OrderedPolymer` `OrderedMonomer` and `OrderedPolymerSpecies` are new data members that represent sequences of parts. A piece of DNA with many binding sites is represented as an OrderedPolymerSpecies. Because mechanisms and Components are agnostic of binding location; that is to say, they usually have code that looks like this: `make_complex([dna,protein])`, we must give them some way of knowing _where_ on that `dna` the `protein` should end up. To do this, you must pass an `OrderedMonomer` into a `Component` which belongs to an `OrderedPolymer`. Through the use of the `Complex()` function, the proper complex will be made, to the correct member of the `OrderedPolymer` which the `OrderedMonomer` belongs to.\n",
    "\n",
    "3. `DNA_part` inherits from `Component`, and every dna part such as `Promoter` now inherits from `DNA_part` instead of `Component`. This allows the storage of things like position, direction, and parent construct inside anything that used to be a `Component`\n",
    "\n",
    "4. `CDS` is a `DNA_part` which gives rise to a `Species`. It felt weird to put `Protein` objects into a DNA sequence so I created this. It only makes the protein if it is read in the 'forward' direction.\n",
    "\n",
    "4. `DNA_construct` contains a list of `DNA_part`s together with their directions. `DNA_construct` also has `update_species()` and `update_reactions()` so that it can generate the necessary species and reactions if it will be part of a CRN. Likewise, `RNA_construct` does the same thing for RNA. A `DNA_construct` can give rise to many `RNA_construct`s when it compiles. This is done through the action of `explore_txtl()`\n",
    "\n",
    "5. `explore_txtl()` is a function of `DNA_construct` which figures out the RNA and Proteins that a DNA produces. It works basically by going along a DNA part by part, remembering which parts came before, and taking into account the directions of parts. In general it will traverse a `DNA_construct` in the \"forward\" and \"reverse\" directions. It is hard coded to know that `RBS` parts make proteins and `Promoter` parts make RNAs, but only when their orientation is \"forward\" relative to the direction it's currently looking at. Other things are taken account too, for example, an `RBS` can only make a `Protein` if it is part of an RNA. Most of this logic is encapsulated in an object called `TxTl_explorer`.\n",
    "\n",
    "6. `update_components()` is a function of `DNA_construct` which creates many copies of the `DNA_part`s which make up the `DNA_construct`. This is necessary for making all combinatorial complexes. First, every `DNA_part` is fed in an `OrderedComplexSpecies` corresponding to the unbound DNA, which has the appropriate \"bindloc_\". Then, the `Species` produced by `update_species()` each have complexes in only one location. All possible combinations of bound species are then generated from this set of `OrderedComplexSpecies`. Then, these combinatorial complexes are fed _back_ into the `DNA_part`s, generating a new `DNA_part` that is responsible for generating species and reactions that include that combinatorial complex. For example, if you have  `DNA_construct` containing a `Promoter` and an `AttachmentSite`, then `update_components()` should generate two `Promoter` parts, one that has the unbound `OrderedComplexSpecies` and one that has an `OrderedComplexSpecies` that has integrases bound to it. Then, these new `Component`s can be used to make species and reactions using their respective `update_species()` and `update_reactions()`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Another complex DNA_construct example, where both DNA and RNA can be bound combinatorially\n",
    "ptet.color=\"forestgreen\"\n",
    "utr1.color = \"tan\"\n",
    "t16.color=\"black\"\n",
    "gfp.color=\"green\"\n",
    "rfp.color=\"red\"\n",
    "polycis_construct = DNA_construct([[t16,\"reverse\"],[cfp,\"reverse\"],[utr1,\"reverse\"],[pconst,\"reverse\"],[ptet,\"forward\"],[utr1,\"forward\"],[gfp,\"forward\"],[utr1,\"forward\"],[rfp,\"forward\"],[t16,\"forward\"]],circular=True)\n",
    "if(dpl_enabled):\n",
    "    plotConstruct(polycis_construct,debug=False,plot_rnas=True,showlabels=[RegulatedPromoter,RBS,CDS,Terminator])\n",
    "\n",
    "\n",
    "#some very basic parameters are defined\n",
    "parameters={\"cooperativity\":2,\"kb\":100, \"ku\":10, \"ktx\":.05, \"ktl\":.2, \"kdeg\":2,\"kint\":.05,\"kdil\":.5}\n",
    "\n",
    "components = [polycis_construct]\n",
    "myMixture = TxTlDilutionMixture(name = \"txtl\", parameters = parameters, components = components)\n",
    "myCRN = myMixture.compile_crn()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "colordict={\"complex\":\"cyan\",\"protein\":\"green\",\n",
    "            \"GFP\":\"lightgreen\",\n",
    "            \"Ribo\":\"pink\",\n",
    "            \"RNAP\":\"skyblue\",\n",
    "            \"RFP\":\"red\",\n",
    "            \"CFP\":\"blue\",\n",
    "            \"ribosome\":\"blue\",\n",
    "            \"dna\":\"white\",\"rna\":\"orange\",\n",
    "            \"ligand\":\"pink\",\"phosphate\":\"yellow\",\"nothing\":\"purple\"}\n",
    "plotNetwork(myCRN,colordict=colordict,iterations=2000,rseed=11,posscale=.2,export=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
